### PR TITLE
feat(user_metadata): add support for user metadata

### PR DESCRIPTION
**Pull request by porcellus**
## Summary of change
Adds user metadata support

## Related issues
- #381 
- #239 

## Test Plan

Added multiple tests

## Documentation changes

Most of it will be done as part of the backend implementation.

## Checklist for important updates
- [x] Changelog has been updated
    - [x] If there are any db schema changes, mention those changes clearly
- [x] `coreDriverInterfaceSupported.json` file has been updated (if needed)
- [x] `pluginInterfaceSupported.json` file has been updated (if needed)
- [x] Changes to the version if needed
   - In `build.gradle`
- [x] Had installed and ran the pre-commit hook
- [x] If there are new dependencies that have been added in `build.gradle`, please make sure to add them in `implementationDependencies.json`.
- [x] Issue this PR against the latest non released version branch.
   - To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
   - If no such branch exists, then create one from the latest released branch.

## Remaining TODOs for this PR
- [ ] Update docs w/ database setup

### CODE REVIEW

File: src/test/java/io/supertokens/test/userMetadata/UserMetadataTest.java [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+        assertNotNull(process.checkOrWaitForEvent(ProcessState.PROCESS_STATE.STOPPED));
+    }
+
+    /**
+     * updates the metadata of a user and merges the result
+     *
+     * @throws Exception
+     */
+    @Test
+    public void updateMetadata() throws Exception {
```
Comment (rishabhpoddar) [inline src/test/java/io/supertokens/test/userMetadata/UserMetadataTest.java:120]: You should also add a test to test concurrent updates to a user's metadata to ensure race conditions are handled correctly.

File: src/main/java/io/supertokens/inmemorydb/queries/UserMetadataQueries.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+        return update(start, QUERY.toString(), pst -> pst.setString(1, userId));
+    }
+
+    public static int setUserMetadata_Transaction(Start start, Connection con, String userId, JsonObject metadata)
+            throws SQLException, StorageQueryException {
+
+        String QUERY = "INSERT INTO " + getConfig(start).getUserMetadataTable()
+                + "(user_id, user_metadata) VALUES(?, ?) "
+                + "ON CONFLICT(user_id) DO UPDATE SET user_metadata=excluded.user_metadata;";
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/inmemorydb/queries/UserMetadataQueries.java:58]: What does `user_metadata=excluded.user_metadata` mean?

File: src/main/java/io/supertokens/inmemorydb/queries/UserMetadataQueries.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    public static JsonObject getUserMetadata(Start start, String userId) throws SQLException, StorageQueryException {
+        String QUERY = "SELECT user_metadata FROM " + getConfig(start).getUserMetadataTable() + " WHERE user_id = ?";
+        return execute(start, QUERY, pst -> pst.setString(1, userId), result -> {
+            if (result.next()) {
+                JsonParser jp = new JsonParser();
+                return jp.parse(result.getString("user_metadata")).getAsJsonObject();
+            }
+            return null;
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/inmemorydb/queries/UserMetadataQueries.java:88]: I think you should return an empty JSON object here instead of checking for `null` in the core, since maybe in the future we might have to query this in other places in the core too, and at that time, we will have to remember to do null checks.

File: src/main/java/io/supertokens/inmemorydb/queries/UserMetadataQueries.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        ((ConnectionWithLocks) con).lock(userId + getConfig(start).getUserMetadataTable());
+
+        String QUERY = "SELECT user_metadata FROM " + getConfig(start).getUserMetadataTable() + " WHERE user_id = ?";
+
+        return execute(con, QUERY, pst -> pst.setString(1, userId), result -> {
+            if (result.next()) {
+                JsonParser jp = new JsonParser();
+                return jp.parse(result.getString("user_metadata")).getAsJsonObject();
+            }
+            return null;
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/inmemorydb/queries/UserMetadataQueries.java:77]: I think you should return an empty JSON object here instead of checking for `null` in the core, since maybe in the future we might have to query this in other places in the core too, and at that time, we will have to remember to do null checks.

File: src/main/java/io/supertokens/storageLayer/StorageLayer.java [hunk 5]
```diff
@@ -169,6 +171,13 @@ public static JWTRecipeStorage getJWTRecipeStorage(Main main) {
         return (JWTRecipeStorage) getInstance(main).storage;
     }
 
+    public static UserMetadataSQLStorage getUserMetadataStorage(Main main) {
+        if (getInstance(main) == null) {
+            throw new QuitProgramException("please call init() before calling getStorageLayer");
+        }
+        return (UserMetadataSQLStorage) getInstance(main).storage;
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storageLayer/StorageLayer.java:178]: You also need to add:
```java
if (getInstance(main).storage.getType() != STORAGE_TYPE.SQL) {
            // we only support SQL for now
            throw new UnsupportedOperationException("");
        }
```

File: src/main/java/io/supertokens/usermetadata/UserMetadata.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        return storage.startTransaction((con) -> {
+            JsonObject originalMetadata = storage.getUserMetadata_Transaction(con, userId);
+
+            final JsonObject updatedMetadata = originalMetadata == null ? new JsonObject() : originalMetadata;
+            metadataUpdate.entrySet().forEach((entry) -> {
+                updatedMetadata.remove(entry.getKey());
+                if (!entry.getValue().isJsonNull()) {
+                    updatedMetadata.add(entry.getKey(), entry.getValue());
+                }
+            });
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/usermetadata/UserMetadata.java:43]: Please extract this into a utils function cause other recipes (like user roles which contains metadata too) will need to use this.

File: coreDriverInterfaceSupported.json [hunk 7]
```diff
@@ -6,6 +6,7 @@
     "2.9",
     "2.10",
     "2.11",
-    "2.12"
+    "2.12",
+    "2.13"
```
Comment (rishabhpoddar) [inline coreDriverInterfaceSupported.json:10]: You also need to update `pluginInterfaceSupported.json`. It is checked in the comment for this PR, but the file change here is not there.

File: src/test/java/io/supertokens/test/userMetadata/UserMetadataTest.java [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+        update.add("cleared", JsonNull.INSTANCE);
+
+        JsonObject updateResult = UserMetadata.updateUserMetadata(process.getProcess(), userId, update);
+
+        JsonObject newMetadata = UserMetadata.getUserMetadata(process.getProcess(), userId);
+
+        assertEquals(updateResult, newMetadata);
+
+        // We removed what we set to null
+        assert (!newMetadata.has("toClear"));
```
Comment (rishabhpoddar) [inline src/test/java/io/supertokens/test/userMetadata/UserMetadataTest.java:156]:
```suggestion
        assert (!newMetadata.has("cleared"));
```

File: src/test/java/io/supertokens/test/userMetadata/UserMetadataTest.java [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+
+        // First we create the original
+        UserMetadata.updateUserMetadata(process.getProcess(), userId, originalMetadata);
+
+        JsonObject update = new JsonObject();
+        JsonObject updateSubObject = new JsonObject();
+        updateSubObject.addProperty("subsubupdate", "subnew!");
+        update.add("testUpdate", updateSubObject);
+        update.addProperty("testNew", "new!");
+        update.add("cleared", JsonNull.INSTANCE);
```
Comment (rishabhpoddar) [inline src/test/java/io/supertokens/test/userMetadata/UserMetadataTest.java:147]: You shold also try by just doing `update.add("cleared", null);`?

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/inmemorydb/queries/UserMetadataQueries.java (hunk 4 above)
Comment (porcellus) [reply src/main/java/io/supertokens/inmemorydb/queries/UserMetadataQueries.java:77]: I think returning null from the DB could be useful in the future (e.g.: if we decide the endpoint should return if this was the first time creating or if it was an update). Also, other places in the core could just query this through the recipe function instead of directly accessing the DB. I figured why throw this info away on the interface level.

File: src/main/java/io/supertokens/inmemorydb/queries/UserMetadataQueries.java (hunk 2 above)
Comment (porcellus) [reply src/main/java/io/supertokens/inmemorydb/queries/UserMetadataQueries.java:58]: it just sets the user_metadata to what would've been inserted if the conflict wasn't there

File: src/test/java/io/supertokens/test/userMetadata/UserMetadataTest.java (hunk 9 above)
Comment (porcellus) [reply src/test/java/io/supertokens/test/userMetadata/UserMetadataTest.java:147]: That does the same thing in effect: the `add` function translates the null into a JsonNull.

File: src/main/java/io/supertokens/inmemorydb/queries/UserMetadataQueries.java (hunk 2 above)
Comment (rishabhpoddar) [reply src/main/java/io/supertokens/inmemorydb/queries/UserMetadataQueries.java:58]: Cool

File: src/main/java/io/supertokens/usermetadata/UserMetadata.java [hunk 10]
```diff
@@ -34,13 +34,8 @@ public static JsonObject updateUserMetadata(Main main, @Nonnull String userId, @
...
 
-            final JsonObject updatedMetadata = originalMetadata == null ? new JsonObject() : originalMetadata;
-            metadataUpdate.entrySet().forEach((entry) -> {
-                updatedMetadata.remove(entry.getKey());
-                if (!entry.getValue().isJsonNull()) {
-                    updatedMetadata.add(entry.getKey(), entry.getValue());
-                }
-            });
+            JsonObject updatedMetadata = originalMetadata == null ? new JsonObject() : originalMetadata;
+            MetadataUtils.shallowMergeMetadataUpdate(updatedMetadata, metadataUpdate);
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/usermetadata/UserMetadata.java:38]: You may want to extrac this function to a global utils file. Since another recipe would be using it. So it would be weird if that recipe will refer directly to Metadata recipe.

File: src/test/java/io/supertokens/test/userMetadata/api/SetUserMetadataAPITest2_13.java [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+        assertEquals(2, resp.entrySet().size());
+        assertEquals("OK", resp.get("status").getAsString());
+        assert (resp.has("metadata"));
+        JsonObject updateResult = resp.getAsJsonObject("metadata");
+
+        JsonObject newMetadata = UserMetadata.getUserMetadata(process.getProcess(), userId);
+        assertEquals(updateResult, newMetadata);
+
+        // We removed what we set to null
+        assert (!newMetadata.has("toClear"));
```
Comment (rishabhpoddar) [inline src/test/java/io/supertokens/test/userMetadata/api/SetUserMetadataAPITest2_13.java:142]:
```suggestion
        assert (!newMetadata.has("cleared"));
```

Comment (rishabhpoddar) [reply src/test/java/io/supertokens/test/userMetadata/api/SetUserMetadataAPITest2_13.java:142]: Please make sure this type of error is fixed in all tests.

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/usermetadata/UserMetadata.java (hunk 10 above)
Comment (porcellus) [reply src/main/java/io/supertokens/usermetadata/UserMetadata.java:39]: I've moved the entire `MetadataUtils` into the `Utils` folder. It's still global this way, but I felt like the `Utils` class was getting crowded.

Comment (rishabhpoddar) [reply src/main/java/io/supertokens/usermetadata/UserMetadata.java:39]: ok

**Review by rishabhpoddar — APPROVED** [review]

File: src/main/java/io/supertokens/usermetadata/UserMetadata.java [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+            throws StorageQueryException, StorageTransactionLogicException {
+        UserMetadataSQLStorage storage = StorageLayer.getUserMetadataStorage(main);
+
+        return storage.startTransaction((con) -> {
+            JsonObject originalMetadata = storage.getUserMetadata_Transaction(con, userId);
+
+            JsonObject updatedMetadata = originalMetadata == null ? new JsonObject() : originalMetadata;
+            MetadataUtils.shallowMergeMetadataUpdate(updatedMetadata, metadataUpdate);
+
+            storage.setUserMetadata_Transaction(con, userId, updatedMetadata);
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/usermetadata/UserMetadata.java:41]: Add explicit testing for if the select statement returns an empty result and by the time it comes here, another thread has already inserted the row. Then what happens here? Does it retry?

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/usermetadata/UserMetadata.java (hunk 12 above)
Comment (porcellus) [reply src/main/java/io/supertokens/usermetadata/UserMetadata.java:41]: Added the test in the Postgres and MySQL PRs.

Comment (rishabhpoddar) [reply src/main/java/io/supertokens/usermetadata/UserMetadata.java:41]: okay

**Review by rishabhpoddar — APPROVED** [review]

**Comment by rishabhpoddar** [comment]
@porcellus please resolve conflicts
